File: headfirst/02_observer/weather_station.go
package main

import (
	"container/list"
	"fmt"
	"reflect"
)

type Subject interface {
	Register(Observer)
	Remove(Observer)
	NotifyAll()
}

type Observer interface {
	Update(temp, humidity, pressure float32)
}

type Displayer interface {
	Display()
}

type WeatherData struct {
	observers   *list.List
	temperature float32
	humidity    float32
	pressure    float32
}

func NewWeatherData() *WeatherData {
	return &WeatherData{observers: list.New()}
}

func (w *WeatherData) Register(o Observer) {
	w.observers.PushBack(o)
}

func (w *WeatherData) Remove(o Observer) {
	elem := w.observers.Front()
	for elem != nil {
		if reflect.DeepEqual(elem.Value, o) {
			w.observers.Remove(elem)
			return
		}
		elem = elem.Next()
	}
}

func (w *WeatherData) NotifyAll() {
	elem := w.observers.Front()
	for elem != nil {
		elem.Value.(Observer).Update(w.temperature, w.humidity, w.pressure)
		elem = elem.Next()
	}
}

func (w *WeatherData) changed() {
	w.NotifyAll()
}

func (w *WeatherData) SetMeasurements(temperature, humidity, pressure float32) {
	w.temperature = temperature
	w.humidity = humidity
	w.pressure = pressure
	w.changed()
}

type CurrentCondition struct {
	temperature, humidity float32
	weatherData           Subject
}

func NewCurrentCondition(weatherData Subject) *CurrentCondition {
	cond := &CurrentCondition{weatherData: weatherData}
	weatherData.Register(cond)
	return cond
}

func (c *CurrentCondition) Update(temperature, humidity, pressure float32) {
	c.temperature = temperature
	c.humidity = humidity
	c.Display()
}

func (c *CurrentCondition) Display() {
	fmt.Printf("Current conditions: %.2f F degrees and %.2f%% humidity\n", c.temperature, c.humidity)
}

type StatisticsCondition struct {
	temperature, humidity float32
	weatherData           Subject
}

func NewStatisticsCondition(weatherData Subject) *StatisticsCondition {
	cond := &StatisticsCondition{weatherData: weatherData}
	weatherData.Register(cond)
	return cond
}

func (s *StatisticsCondition) Update(temperature, humidity, pressure float32) {
	s.temperature = temperature
	s.humidity = humidity
	s.Display()
}

func (s *StatisticsCondition) Display() {
	fmt.Println("Statistics conditions: 🤡")
}

type ForecastCondition struct {
	temperature, humidity float32
	weatherData           Subject
}

func NewForecastCondition(weatherData Subject) *ForecastCondition {
	cond := &ForecastCondition{weatherData: weatherData}
	weatherData.Register(cond)
	return cond
}

func (f *ForecastCondition) Update(temperature, humidity, pressure float32) {
	f.temperature = temperature
	f.humidity = humidity
	f.Display()
}

func (f *ForecastCondition) Display() {
	fmt.Println("Forecast conditions: 😈")
}

func main() {
	weatherData := NewWeatherData()

	curCond := NewCurrentCondition(weatherData)
	statCond := NewStatisticsCondition(weatherData)
	foreCond := NewForecastCondition(weatherData)

	_ = curCond
	_ = statCond
	_ = foreCond

	weatherData.SetMeasurements(80, 65, 30.4)
	weatherData.SetMeasurements(82, 70, 29.2)
	weatherData.SetMeasurements(78, 90, 29.2)

	fmt.Println("================== remove statCond")
	weatherData.Remove(statCond)
	weatherData.SetMeasurements(80, 65, 30.4)

	fmt.Println("==================remove foreCond")
	weatherData.Remove(foreCond)
	weatherData.SetMeasurements(80, 65, 30.4)

	fmt.Println("==================remove curCond")
	weatherData.Remove(curCond)
	weatherData.SetMeasurements(80, 65, 30.4)

}

File: headfirst/02_observer/WeatherStation.java
import java.util.ArrayList;


interface Subject {
    void registerObserver(Observer o);
    void removeObserver(Observer o);
    void notifyObservers();
}

interface Observer{
    void update(float temp, float humidity, float pressure);
}

interface DisplayElement {
    void display();
}

class WeatherData implements Subject {
    private ArrayList<Observer> observers;
    private float temperature;
    private float humidity;
    private float pressure;

    WeatherData() {
        observers = new ArrayList<>();
    }

    public void registerObserver(Observer o) {
        observers.add(o);
    }

    public void removeObserver(Observer o) {
        int i = observers.indexOf(o);
        if (i > 0) {
            observers.remove(i);
        }
    }

    public void notifyObservers() {
        for (int i=0; i < observers.size(); i++) {
            Observer observer = observers.get(i);
            observer.update(temperature, humidity, pressure);
        }
    }

    public void measurementsChanged() {
        notifyObservers();
    }

    public void setMeasurements(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;
        measurementsChanged();
    }

}

class CurrentConditionsDisplay implements Observer, DisplayElement {
    private float temperature;
    private float humidity;
    private Subject weatherData;

    CurrentConditionsDisplay(Subject weatherData) {
        this.weatherData = weatherData;
        weatherData.registerObserver(this);
    }

    public void update(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        display();
    }

    public void display() {
        System.out.println("Current conditions: " + temperature
            + "F degrees and " + humidity + "% humidity");  
    }
}

class StatisticsDisplay implements Observer, DisplayElement {
    private float temperature;
    private float humidity;
    private Subject weatherData;

    StatisticsDisplay(Subject weatherData) {
        this.weatherData = weatherData;
        weatherData.registerObserver(this);
    }

    public void update(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        display();
    }

    public void display() {
        System.out.println("Statistics conditions:  😀");  
    }
}

class ForecastDisplay implements Observer, DisplayElement {
    private float temperature;
    private float humidity;
    private Subject weatherData;

    ForecastDisplay(Subject weatherData) {
        this.weatherData = weatherData;
        weatherData.registerObserver(this);
    }

    public void update(float temperature, float humidity, float pressure) {
        this.temperature = temperature;
        this.humidity = humidity;
        display();
    }

    public void display() {
        System.out.println("Forecast conditions:  🤡");  
    }
}


public class WeatherStation {
    public static void main(String[] args) {
        WeatherData weatherData = new WeatherData();

        CurrentConditionsDisplay currentDisplay = new CurrentConditionsDisplay(weatherData);
        StatisticsDisplay statisticsDisplay = new StatisticsDisplay(weatherData);
        ForecastDisplay forecastDisplay = new ForecastDisplay(weatherData);

        weatherData.setMeasurements(80, 65, 30.4f);
        weatherData.setMeasurements(82, 70, 29.2f);
        weatherData.setMeasurements(78, 90, 29.2f);
    }
}